eth/protocols/snap/handler.go

```diff
@@ -0,0 +1,514 @@
+	// tries, but proofs can only be attached for the last one.
+	OnStorage(peer *Peer, id uint64, hashes [][]common.Hash, slots [][][]byte, proof [][]byte) error
+
+	// OnByteCodes is a callback method to invoke when a batch of contract
+	// bytes codes are received from a remote peer.
+	OnByteCodes(peer *Peer, id uint64, bytecodes [][]byte) error
+
+	// OnTrieNodes is a callback method to invoke when a batch of trie nodes
+	// are received from a remote peer.
+	OnTrieNodes(peer *Peer, id uint64, nodes [][]byte) error
+}
+
+// MakeProtocols constructs the P2P protocol definitions for `snap`.
+func MakeProtocols(backend Backend, dnsdisc enode.Iterator) []p2p.Protocol {
+	protocols := make([]p2p.Protocol, len(protocolVersions))
+	for i, version := range protocolVersions {
+		version := version // Closure
+
+		protocols[i] = p2p.Protocol{
+			Name:    protocolName,
+			Version: version,
+			Length:  protocolLengths[version],
+			Run: func(p *p2p.Peer, rw p2p.MsgReadWriter) error {
+				return backend.RunPeer(newPeer(version, p, rw), func(peer *Peer) error {
+					return handle(backend, peer)
+				})
+			},
+			NodeInfo: func() interface{} {
+				return nodeInfo(backend.Chain())
+			},
+			PeerInfo: func(id enode.ID) interface{} {
+				return backend.PeerInfo(id)
+			},
+			Attributes:     []enr.Entry{&enrEntry{}},
+			DialCandidates: dnsdisc,
+		}
+	}
+	return protocols
+}
+
+// handle is the callback invoked to manage the life cycle of a `snap` peer.
+// When this function terminates, the peer is disconnected.
+func handle(backend Backend, peer *Peer) error {
+	for {
+		if err := handleMessage(backend, peer); err != nil {
+			peer.Log().Debug("Message handling failed in `snap`", "err", err)
+			return err
+		}
+	}
+}
+
+// handleMessage is invoked whenever an inbound message is received from a
+// remote peer on the `spap` protocol. The remote connection is torn down upon
+// returning any error.
+func handleMessage(backend Backend, peer *Peer) error {
+	// Read the next message from the remote peer, and ensure it's fully consumed
+	msg, err := peer.rw.ReadMsg()
+	if err != nil {
+		return err
+	}
+	if msg.Size > maxMessageSize {
+		return fmt.Errorf("%w: %v > %v", errMsgTooLarge, msg.Size, maxMessageSize)
+	}
+	defer msg.Discard()
+
+	// Handle the message depending on its contents
+	switch {
+	case msg.Code == getAccountRangeMsg:
+		// Decode the account retrieval request
+		var req getAccountRangeData
+		if err := msg.Decode(&req); err != nil {
+			return fmt.Errorf("%w: message %v: %v", errDecode, msg, err)
+		}
+		if req.Bytes > softResponseLimit {
+			req.Bytes = softResponseLimit
+		}
+		// Retrieve the requested state and bail out if non existent
+		tr, err := trie.New(req.Root, backend.Chain().StateCache().TrieDB())
+		if err != nil {
+			return p2p.Send(peer.rw, accountRangeMsg, &accountRangeData{ID: req.ID})
+		}
+		it, err := backend.Chain().Snapshot().AccountIterator(req.Root, req.Origin)
```

Comment:
Scenario:  some peer is fetching `snap` data from me -- so it's got snap proto enabled, but does not yet have a snapshot (it doesn't even have a full trie yet). 

At this time, if I send it a `getAccountRangeMsg`, I think ` backend.Chain().Snapshot()` will return `nil`, and it will crash here.

Reply:
The `Snapshot` method returns the entire snapshot tree, which will never be nil if snapshots are enabled (it's content might be empty, but that's a different story).